src/ir/principal-type.cpp

```diff
@@ -0,0 +1,1033 @@
+/*
+ * Copyright 2025 WebAssembly Community Group participants
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <algorithm>
+#include <iostream>
+
+#include "principal-type.h"
+#include "support/disjoint_sets.h"
+
+namespace wasm {
+
+namespace {
+
+#ifndef NDEBUG
+
+bool valid(const VarAbsHeapType& type) {
+  if (!type.ht.isBasic()) {
+    return false;
+  }
+  // Sharedness is represented in `share`, not `ht`.
```

Comment:
Are these validity constraints described in the header?

Reply:
I've added comments for them.